planex/spec.py

```diff
@@ -365,18 +367,65 @@ def specpath(self):
         """Return the path to the spec file"""
         return self.path
 
-    def expand_patchqueues(self):
+    def rewrite_spec(self):
         """
-        Insert the names of all patchqueue patches into the spec file
+        Rewrite the sources and patches in the spec file, also inserting the
+        names of all patchqueue patches into it
         """
-        patchqueues = [self._patchqueues[key] for key
-                       in sorted(self._patchqueues.keys())]
-        if not patchqueues:
-            return self.spectext
+        if self._patches or self._patchqueues:
+            planex.patchqueue.check_spec_supports_patchqueues(self)
 
-        series = sum([pq.series() for pq in patchqueues], [])
-        spec = planex.patchqueue.expand_patchqueue(self, series)
-        return "".join(list(spec))
+        def is_source_or_patch_line(line):
+            """True if the line does start with Source or Patch"""
+            source = re.compile("^source", re.IGNORECASE)
+            patch = re.compile("^patch", re.IGNORECASE)
+            return bool(patch.match(line) or source.match(line))
+
+        def first_index(iterable, predicate):
+            """
+            Returns the index of first value in the iterable
+            for which predicate(value) is true.
+            """
+            return next(i for i, v in enumerate(iterable) if predicate(v))
+
+        split_at = first_index(self.spectext, is_source_or_patch_line)
```

Comment:
This assumes that there is at least a `Source` or `Patch` line in the spec file, but I thought that figuring out how to support specfiles not having sources or patches was not worth the effort

Reply:
This will probably fail with the `vendor-drivers` spec file then. We can add a fake source if necessary.

Reply:
The old code was not able to deal with absent sources, but it was not rewriting them either. I think we should add the fake sources, I am not sure how to decide where to put them otherwise

Reply:
Actually, if there are no sources or patches, we could easily fix the code

Reply:
Done!